WasbHook test_sas_token_connection fails with azure-storage-blob==12.30.0

I have a PR (#68479) failing on the **Compat 2.11.1** CI job on tests that have nothing to do with Azure. Looking at the logs, the actual failures are all in `test_wasb.py`:

```
FAILED providers/microsoft/azure/tests/unit/microsoft/azure/hooks/test_wasb.py::TestWasbHook::test_sas_token_connection[sas_conn_id-sas_token]
FAILED providers/microsoft/azure/tests/unit/microsoft/azure/hooks/test_wasb.py::TestWasbHook::test_sas_token_connection[extra__wasb__sas_conn_id-extra__wasb__sas_token]
FAILED providers/microsoft/azure/tests/unit/microsoft/azure/hooks/test_wasb.py::TestWasbHook::test_sas_token_connection[http_sas_conn_id-sas_token]
FAILED providers/microsoft/azure/tests/unit/microsoft/azure/hooks/test_wasb.py::TestWasbHook::test_sas_token_connection[extra__wasb__http_sas_conn_id-extra__wasb__sas_token]
```

Error:

```
AssertionError: assert False
  where False = 'https://login.blob.core.windows.net/'.endswith('token/')
  where 'https://login.blob.core.windows.net/' = BlobServiceClient(...).url
```

The test asserts `conn.url.endswith(sas_token + "/")` but `BlobServiceClient.url` now returns `https://login.blob.core.windows.net/` instead of the URL with the SAS token appended.

Digging into it, this seems to be caused by `azure-storage-blob` being bumped from **12.29.0 → 12.30.0** in the constraints update `e3ed4035ee` (2026-06-12). The Compat test runs with `--providers-skip-constraints` so it picks up 12.30.0 from PyPI, and in that version `BlobServiceClient.url` appears to no longer reflect the SAS token in the URL.

This is blocking unrelated PRs. Can someone take a look?

--- a/providers/microsoft/azure/src/airflow/providers/microsoft/azure/hooks/wasb.py
+++ b/providers/microsoft/azure/src/airflow/providers/microsoft/azure/hooks/wasb.py
@@ -30,6 +30,7 @@ import logging
 import os
 from typing import TYPE_CHECKING, Any, cast
 
+from azure.core.credentials import AzureSasCredential
 from azure.core.exceptions import HttpResponseError, ResourceExistsError, ResourceNotFoundError
 from azure.identity import ClientSecretCredential
 from azure.identity.aio import (
@@ -203,7 +204,9 @@ class WasbHook(BaseHook):
         if sas_token:
             if sas_token.startswith("https"):
                 return BlobServiceClient(account_url=sas_token, **extra)
-            return BlobServiceClient(account_url=f"{account_url.rstrip('/')}/{sas_token}", **extra)
+            return BlobServiceClient(
+                account_url=account_url, credential=AzureSasCredential(sas_token), **extra
+            )
 
         # Fall back to old auth (password) or use managed identity if not provided.
         credential: str | TokenCredential | None = conn.password
@@ -671,7 +674,7 @@ class WasbAsyncHook(WasbHook):
                 self.blob_service_client = AsyncBlobServiceClient(account_url=sas_token, **extra)
             else:
                 self.blob_service_client = AsyncBlobServiceClient(
-                    account_url=f"{account_url.rstrip('/')}/{sas_token}", **extra
+                    account_url=account_url, credential=AzureSasCredential(sas_token), **extra
                 )
             return self.blob_service_client
 

--- a/providers/microsoft/azure/tests/unit/microsoft/azure/hooks/test_wasb.py
+++ b/providers/microsoft/azure/tests/unit/microsoft/azure/hooks/test_wasb.py
@@ -45,6 +45,13 @@ CONN_STRING = (
 ACCESS_KEY_STRING = "AccountName=name;skdkskd"
 PROXIES = {"http": "http_proxy_uri", "https": "https_proxy_uri"}
 
+# A SAS token is a query string (not a path segment). Use a representative token and the
+# equivalent full SAS URL so the resulting BlobServiceClient.url carries it in the query.
+# This is what azure-storage-blob actually uses to sign requests, and is stable across SDK
+# versions.
+SAS_TOKEN = "?sv=2021-08-06&ss=b&srt=co&sp=r&sig=samplesignature"
+HTTPS_SAS_TOKEN = f"https://login.blob.core.windows.net/{SAS_TOKEN}"
+
 
 @pytest.fixture
 def mocked_blob_service_client():
@@ -147,24 +154,24 @@ class TestWasbHook:
                 conn_id="sas_conn_id",
                 conn_type=self.connection_type,
                 login=self.login,
-                extra={"sas_token": "token", "proxies": self.proxies},
+                extra={"sas_token": SAS_TOKEN, "proxies": self.proxies},
             ),
             Connection(
                 conn_id=self.extra__wasb__sas_conn_id,
                 conn_type=self.connection_type,
                 login=self.login,
-                extra={"extra__wasb__sas_token": "token", "proxies": self.proxies},
+                extra={"extra__wasb__sas_token": SAS_TOKEN, "proxies": self.proxies},
             ),
             Connection(
                 conn_id=self.http_sas_conn_id,
                 conn_type=self.connection_type,
-                extra={"sas_token": "https://login.blob.core.windows.net/token", "proxies": self.proxies},
+                extra={"sas_token": HTTPS_SAS_TOKEN, "proxies": self.proxies},
             ),
             Connection(
                 conn_id=self.extra__wasb__http_sas_conn_id,
                 conn_type=self.connection_type,
                 extra={
-                    "extra__wasb__sas_token": "https://login.blob.core.windows.net/token",
+                    "extra__wasb__sas_token": HTTPS_SAS_TOKEN,
                     "proxies": self.proxies,
                 },
             ),
@@ -364,7 +371,9 @@ class TestWasbHook:
         assert conn.url.startswith("https://")
         if hook_conn.login:
             assert hook_conn.login in conn.url
-        assert conn.url.endswith(sas_token + "/")
+        # The SAS token must be carried in the URL query string so the SDK signs requests with
+        # it.
+        assert sas_token.lstrip("?") in conn.url
 
     @pytest.mark.parametrize(
         argnames="conn_id_str",

--- a/providers/microsoft/azure/tests/unit/microsoft/azure/hooks/test_wasb.py
+++ b/providers/microsoft/azure/tests/unit/microsoft/azure/hooks/test_wasb.py
@@ -374,7 +374,7 @@ class TestWasbHook:
         if sas_token.startswith("https"):
             # HTTPS SAS URL: the full URL is passed to BlobServiceClient as-is;
             # credential handling is left to the SDK.
-            assert conn.url.startswith("https://")
+            assert sas_token in conn.url
         else:
             # Plain SAS token: the hook wraps it in AzureSasCredential.
             assert isinstance(conn.credential, AzureSasCredential)

--- a/providers/microsoft/azure/tests/unit/microsoft/azure/hooks/test_wasb.py
+++ b/providers/microsoft/azure/tests/unit/microsoft/azure/hooks/test_wasb.py
@@ -23,6 +23,7 @@ from unittest import mock
 from unittest.mock import create_autospec
 
 import pytest
+from azure.core.credentials import AzureSasCredential
 from azure.core.exceptions import ResourceNotFoundError
 from azure.storage.blob import BlobServiceClient, ContainerClient
 from azure.storage.blob._models import BlobProperties
@@ -45,12 +46,10 @@ CONN_STRING = (
 ACCESS_KEY_STRING = "AccountName=name;skdkskd"
 PROXIES = {"http": "http_proxy_uri", "https": "https_proxy_uri"}
 
-# A SAS token is a query string (not a path segment). Use a representative token and the
-# equivalent full SAS URL so the resulting BlobServiceClient.url carries it in the query.
-# This is what azure-storage-blob actually uses to sign requests, and is stable across SDK
-# versions.
-SAS_TOKEN = "?sv=2021-08-06&ss=b&srt=co&sp=r&sig=samplesignature"
-HTTPS_SAS_TOKEN = f"https://login.blob.core.windows.net/{SAS_TOKEN}"
+# https://learn.microsoft.com/en-us/rest/api/storageservices/create-service-sas
+# The hook wraps plain SAS tokens in AzureSasCredential for the SDK to sign requests.
+SAS_TOKEN = "sv=2021-08-06&ss=b&srt=sco&sp=r&sig=samplesignature"
+HTTPS_SAS_TOKEN = f"https://login.blob.core.windows.net/?{SAS_TOKEN}"
 
 
 @pytest.fixture
@@ -311,8 +310,12 @@ class TestWasbHook:
         self, mocked_connection, mocked_blob_service_client
     ):
         WasbHook(wasb_conn_id="testconn").get_conn()
+        called_credential = mocked_blob_service_client.call_args[1]["credential"]
+        assert isinstance(called_credential, AzureSasCredential)
+        assert called_credential.signature == "SAStoken"
         mocked_blob_service_client.assert_called_once_with(
-            account_url="https://testaccountname.blob.core.windows.net/SAStoken",
+            account_url="https://testaccountname.blob.core.windows.net/",
+            credential=called_credential,
             sas_token="SAStoken",
         )
 
@@ -368,12 +371,14 @@ class TestWasbHook:
         hook_conn = hook.get_connection(hook.conn_id)
         sas_token = hook_conn.extra_dejson[extra_key]
         assert isinstance(conn, BlobServiceClient)
-        assert conn.url.startswith("https://")
-        if hook_conn.login:
-            assert hook_conn.login in conn.url
-        # The SAS token must be carried in the URL query string so the SDK signs requests with
-        # it.
-        assert sas_token.lstrip("?") in conn.url
+        if sas_token.startswith("https"):
+            # HTTPS SAS URL: the full URL is passed to BlobServiceClient as-is;
+            # credential handling is left to the SDK.
+            assert conn.url.startswith("https://")
+        else:
+            # Plain SAS token: the hook wraps it in AzureSasCredential.
+            assert isinstance(conn.credential, AzureSasCredential)
+            assert conn.credential.signature == sas_token
 
     @pytest.mark.parametrize(
         argnames="conn_id_str",

